sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2529,19 +2530,19 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
      */
     private def resolveSubQueries(plan: LogicalPlan, outer: LogicalPlan): LogicalPlan = {
       plan.transformAllExpressionsWithPruning(_.containsPattern(PLAN_EXPRESSION), ruleId) {
-        case s @ ScalarSubquery(sub, _, exprId, _, _, _) if !sub.resolved =>
+        case s @ ScalarSubquery(sub, _, exprId, _, _, _) if !sub.analyzed =>
```

Comment:
OK we can't rely on the `analyzed` flag here. Another idea: let's use `TreeNodeTag`. In `resolveSubQuery`, we attach a `TreeNodeTag` to the resolved subquery, then here we can skip resolving it again if the `TreeNodeTag` is present.

Reply:
Thanks for your help, I will keep trying.